paimon-core/src/main/java/org/apache/paimon/tag/TagAutoCreation.java

```diff
@@ -144,29 +157,66 @@ private void tryToTag(Snapshot snapshot) {
+            tagManager.createTag(snapshot, tagName, timeRetained, callbacks);
             nextTag = periodHandler.nextTagTime(thisTag);
+        }
+    }
+
+    private void expireTags() {
+        Set<String> deleteTags = new HashSet<>();
+        deleteTags.addAll(getExpireTagsByNumRetainedMax());
+        deleteTags.addAll(getExpireTagsByTimeRetained());
+        deleteTags.forEach(
+                tag -> tagManager.deleteTag(tag, tagDeletion, snapshotManager, callbacks));
+    }
 
-            if (numRetainedMax != null) {
-                // only handle auto-created tags here
-                SortedMap<Snapshot, List<String>> tags = tagManager.tags(periodHandler::isAutoTag);
-                if (tags.size() > numRetainedMax) {
-                    int toDelete = tags.size() - numRetainedMax;
-                    int i = 0;
-                    for (List<String> tag : tags.values()) {
-                        tagManager.deleteTag(
-                                checkAndGetOneAutoTag(tag),
-                                tagDeletion,
-                                snapshotManager,
-                                callbacks);
-                        i++;
-                        if (i == toDelete) {
-                            break;
-                        }
+    private Set<String> getExpireTagsByNumRetainedMax() {
+        Set<String> deleteTags = new HashSet<>();
+        if (numRetainedMax != null) {
+            // only handle auto-created tags here
+            SortedMap<Snapshot, List<String>> tags = tagManager.tags(periodHandler::isAutoTag);
+            if (tags.size() > numRetainedMax) {
+                int toDelete = tags.size() - numRetainedMax;
+                int i = 0;
+                for (List<String> tag : tags.values()) {
+                    String tagName = checkAndGetOneAutoTag(tag);
+                    LOG.info(
+                            "Delete tag {}, because the number of auto-created tags reached numRetainedMax of {}.",
+                            tagName,
+                            numRetainedMax);
+                    deleteTags.add(tagName);
+                    i++;
+                    if (i == toDelete) {
+                        break;
                     }
                 }
             }
         }
+        return deleteTags;
+    }
+
+    private Set<String> getExpireTagsByTimeRetained() {
+        // handle auto-created and non-auto-created-tags here
+        Set<String> deleteTags = new HashSet<>();
+        SortedMap<Snapshot, List<String>> tags = tagManager.tags();
+        for (Map.Entry<Snapshot, List<String>> entry : tags.entrySet()) {
+            checkArgument(
+                    entry.getKey() instanceof Tag,
+                    "There is a bug, snapshot in tagManager.tags() must be Tag.");
+            Tag tag = (Tag) entry.getKey();
+            LocalDateTime createTime = tag.createTime();
+            Duration timeRetained = tag.timeRetained();
```

Comment:
So what happen if `timeRetained` is null?

Reply:
done
In order to compatibility old version.
New Fields tagCreateTime and tagTimeRetained in Tag should use default value when old version snapshot/tag json to new version Tag Object.